freedomofpress/securedrop-proxy
# Restructures the code base with more object methods

Moves configuration related code and also callbacks as proper methods into Proxy class.

>`proxy.py`

    Now, Proxy class gets a must conf_path argument, and it creates the inital
    `conf` attribute from that.  Proxy also has default on_save and on_done and
    err_on_done methods.  def handle_response(self) method has a `assert
    self.res` to mark that res is populated before this call. This is for mypy

>`main.py`

    Mostly has changes from black

>`entrypoint.py`

    No need for fancy dynamic err_call_back, the proxy object will call
    self.err_call_back if any issue in reading configuration.

The test cases now have their own configuration files to create the proxy
object. Also, to do proper dynamic attachment of any method of Proxy class
we are using https://docs.python.org/3/library/types.html#types.MethodType
so that our own on_save or on_done or err_on_done will be called during
tests.

stretch test will fail as it does not have any variable level type annotation.

## Testing

1. Build the `securedrop-proxy` source distribution:
    - check out this branch
    - run `python setup.py sdist`
2. Build the `.deb`:
    - in a clone of `securedrop-debian-packaging`, run `PKG_PATH=/home/user/src/securedrop-proxy/dist/securedrop-proxy-0.1.6.tar.gz PKG_VERSION=0.1.6 make securedrop-proxy` (adjust the path to your working copy of this repo as necessary)
3. Install the `.deb` in `sd-proxy-buster-template`:
    - `qvm-copy ~/debbuild/packaging/securedrop-proxy_0.1.6+buster_all.deb` and specify `sd-proxy-buster-template` as the target VM
    - In a shell on `sd-proxy-buster-template`, run:
        - `cp ~/QubesIncoming/sd-dev/securedrop-proxy_0.1.6+buster_all.deb /tmp`
        - `sudo apt install /tmp/securedrop-proxy_0.1.6+buster_all.deb`
4. Reboot.
5. Start the SecureDrop client in `sd-svs` and confirm that the proxy is communicating with the server: submissions and replies should populate with no errors.
